### PR TITLE
OPS-2361 Add command diff

### PR DESCRIPTION
**Pull request by alex-harvey-z3q**
commit c6a25faf952e1fbf751984d05a5c27831dedbf92 (HEAD -> feature/OPS-2361-better-diffs-v4, origin/feature/OPS-2361-better-diffs-v4)
Author: Alex Harvey <alex.harvey@cmd.com.au>
Date:   Sun Jul 4 19:17:28 2021 +1000

    Add support for two difflibs
    
    This adds a feature -D, --difflib allowing the user to choose their diff
    library. At the moment, difflib is the default, and dictdiffer can also
    be selected.

commit 1c6eba06494fa1e980d280cda9a32ba820450ab6
Author: Alex Harvey <alex.harvey@cmd.com.au>
Date:   Sun Jul 4 18:43:57 2021 +1000

    Add unified diff version

commit 8377756b7fefc62f881925639a87f5d9e7f9e397
Author: Alex Harvey <alex.harvey@cmd.com.au>
Date:   Sun May 16 07:10:20 2021 +1000

    OPS-2361 Add command diff
    
    This implements a command diff that diffs a running stack's template via
    the fetch_remote_template command and the template generated by the stack.

### CODE REVIEW

File: requirements/dev.txt [hunk 1]
```diff
@@ -22,3 +22,4 @@ tox>=3.23.0,<4.0.0
 troposphere>=2.0.0,<2.1.0
 twine>=1.12.1,<2.0.0
 wheel==0.32.3
+ipdb==0.13.7
```
Comment (mrowlingfox) [inline requirements/dev.txt:25]: Added by mistake?

File: sceptre/plan/actions.py [hunk 2]
```diff
@@ -595,6 +599,43 @@ def fetch_remote_template(self):
...
+
+        elif differ == "dictdiffer":
+            remote_template = yaml.load(remote_template_stream,
+                                        Loader=yaml.BaseLoader)
+            local_template = yaml.load(local_template_stream,
+                                       Loader=yaml.BaseLoader)
+
+            diffs = dictdiffer.diff(remote_template, local_template)
+
+            return_val = list(diffs)
```
Comment (mrowlingfox) [inline sceptre/plan/actions.py]: Split these into separate functions

**Comment by mrowlingfox** [comment]
Can you pleas show some examples of this working

File: requirements/dev.txt (hunk 1 above)
Comment (alex-harvey-z3q) [reply requirements/dev.txt:25]: @mrowlingfox No, I just thought it's convenient to have the debugger available in dev.txt. It's not a must-have obviously.

**Comment by alex-harvey-z3q** [comment]
Examples of the usage after integration with [this](https://github.com/fsa-streamotion/streamotion-platform-platform-sceptre-fx/pull/9) version of `sceptre-fx`:

## help message

Screen Shot 2021-07-06 at 9 51 28 pm

## Test set up

I apply this diff to the network stack:

```diff
--- a/network/templates/network.template.yaml.j2
+++ b/network/templates/network.template.yaml.j2
@@ -31,6 +31,9 @@ Resources:
         -
           Key: Name
           Value: !Ref NameTagValue
+        -
+          Key: Foo
+          Value: Bar
 
   InternetGatewayAttachment:
     Type: AWS::EC2::VPCGatewayAttachment
```

## diff command (default behaviour)

I run:

```text
sceptre-fx /env/network.yaml diff
```

Screen Shot 2021-07-06 at 9 50 36 pm

## diff command (-D)

I run:

```text
sceptre-fx -D /env/network.yaml diff
```

Screen Shot 2021-07-06 at 9 51 14 pm

**Review by mrowlingfox — APPROVED** [review]

**Pull request merged**